src/maps/google-maps-provider.js

```diff
@@ -0,0 +1,58 @@
+/**
+ * @typedef {import('../../types/src/maps/maps.service').MapsAPIProvider} MapsAPIProvider
+ * @typedef {import('../../types/src/maps/maps.service').GetDirectionInfo} GetDirectionInfo
+ * @typedef {import('../../types/src/maps/google-maps-provider').GoogleDirectionResponse} GoogleDirectionResponse
+ * @typedef {import('../../types/src/maps/maps.service').MapsAPIProviderDeps} MapsAPIProviderDeps
+ * @typedef {import('../../types/src/maps/maps.service').DirectionInfoParams} DirectionInfoParams
+ */
+
+import { partial } from '@oldbros/shiftjs';
+import { makeHttpRequest } from '../utils/http-request.utils.js';
+
+const API_URL = 'https://maps.googleapis.com/maps/api/directions/json';
+const GOOGLE_API_KEY = process.env.GOOGLE_API_KEY;
+/**
+ * @param {MapsAPIProviderDeps} deps
+ * @param {DirectionInfoParams} params
+ */
+export const getDirectionInfo = async (deps, { departurePoint, destinationPoint }) => {
+  const { logger } = deps;
+  try {
+    /** @type {GoogleDirectionResponse} */
+    // eslint-disable-next-line max-len
+    const response = await makeHttpRequest(`${API_URL}?origin=${departurePoint.long}, ${departurePoint.lat}&destination=${destinationPoint.long}, ${destinationPoint.lat}&key=${GOOGLE_API_KEY}`);
+    if (!response || response.status !== 'OK') {
+      logger.error(response.status);
+      return null;
+    }
+    const [bestRoute] = response.routes;
+    if (!bestRoute) return null;
+    const [routeInfo] = bestRoute.legs;
+
+    return routeInfo
+      ? {
+        destinationAddress: routeInfo.end_address,
+        departureAddress: routeInfo.start_address,
+        distance: routeInfo.distance.value,
+        duration: routeInfo.duration.value,
+      } : null;
+  } catch (e) {
+    logger.error(e);
+
+    return null;
+  }
+};
```

Comment:
Why do you prefer returning null instead of throwing Error and handling it outside?

Reply:
It is fine to have such an approach sometimes. There are many options for error handling.
 1. Throw errors and process on top - multiple error messages, better logging
 2. Return null or undefined if error - requires a check on top and you have only one error message.
 3. Return { error, data } or [error, data] - contract very similar to a callback, requires a check on top, multiple error messages, no need to try catch on top.
 4. Monad Either - Haskel or Scala way to handle errors with I/O operations in a functional style.